src/server/service/debug.rs

```diff
@@ -292,4 +302,70 @@ impl debugpb_grpc::Debug for Service {
 
         self.handle_response(ctx, sink, f, TAG);
     }
+
+    fn region_consistent_check(
+        &self,
+        ctx: RpcContext,
+        req: RegionConsistentCheckRequest,
+        sink: UnarySink<RegionConsistentCheckResponse>,
+    ) {
+        let region_id = req.get_region_id();
+        let debugger = self.debugger.clone();
+        let router1 = self.raft_router.clone();
+        let router2 = self.raft_router.clone();
+
+        let consistent_check = move || {
+            future::result(debugger.get_store_id())
+                .and_then(move |store_id| region_detail(router2, region_id, store_id))
+                .flatten()
+                .and_then(|detail| consistent_check(router1, detail))
+        };
+        let f = self.pool.spawn_fn(consistent_check).map(|_| {
+            let mut resp = RegionConsistentCheckResponse::new();
+            resp.set_success(true);
+            resp
+        });
+        self.handle_response(ctx, sink, f, "region_consistent_check");
+    }
+}
+
+fn region_detail<T: RaftStoreRouter>(
+    raft_router: T,
+    region_id: u64,
+    store_id: u64,
+) -> Result<impl Future<Item = RegionDetailResponse, Error = Error>, Error> {
+    let mut header = RaftRequestHeader::new();
+    header.set_region_id(region_id);
+    header.mut_peer().set_store_id(store_id);
+    let mut status_request = StatusRequest::new();
+    status_request.set_cmd_type(StatusCmdType::RegionDetail);
+    let mut raft_cmd = RaftCmdRequest::new();
+    raft_cmd.set_header(header);
+    raft_cmd.set_status_request(status_request);
+
+    let (tx, rx) = oneshot::channel();
+    let cb = Callback::Read(box |resp| tx.send(resp).unwrap());
+    raft_router
+        .send_command(raft_cmd, cb)
+        .map(|_| rx.map(|mut r| r.response.take_status_response().take_region_detail()))
+        .map(|rx| rx.map_err(|e| Error::Other(box e)))
+        .map_err(|e| Error::Other(box e))
+}
+
+fn consistent_check<T>(raft_router: T, mut detail: RegionDetailResponse) -> Result<(), Error>
+where
+    T: RaftStoreRouter,
+{
+    let mut header = RaftRequestHeader::new();
+    header.set_region_id(detail.get_region().get_id());
+    header.set_peer(detail.take_leader());
```

Comment:
why use leader here? Here I guess it is the local peer ID.

Reply:
This command is only avaliable for leader. So here the leader should be the peer itself. Otherwise the last propose should fail.

Reply:
If we send the command to peer 2, but here we set the peer with leader 1, I guess you can return the error here directly if the peer is not the leader.

Reply:
If the peer is not leader, it will return error when we call `region_detail` so that we won't call this function any more.

Reply:
I don't think so. `region_detail` will return an empty `RegionDetailResponse`.

Reply:
The conclusion is:
1. If we send `RegionDetail` proposal to **stores which are not leader**, they can return result correctly;
2. We need to return directly if the store is not leader.

In latest commits I have fix it.

Reply:
any update?

Reply:
I still think you should use the local peer id, not leader id here. region_detail won't check leader.

Reply:
Yes, it's already fixed.